File: app/src/main/java/arunkbabu90/popmovies/adapter/ReviewAdapter.java
package arunkbabu90.popmovies.adapter;

import android.support.annotation.NonNull;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import java.util.List;

import arunkbabu90.popmovies.R;
import arunkbabu90.popmovies.Reviews;

public class ReviewAdapter extends RecyclerView.Adapter<ReviewAdapter.ReviewAdapterViewHolder>
{
    private List<Reviews> mReviewList;

    public ReviewAdapter() {
        // Required public constructor
    }

    public void setReviews(List<Reviews> reviewList) {
        mReviewList = reviewList;
        notifyDataSetChanged();
    }

    @NonNull
    @Override
    public ReviewAdapterViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        LayoutInflater inflater = LayoutInflater.from(parent.getContext());
        View view = inflater.inflate(R.layout.item_review, parent, false);
        return new ReviewAdapterViewHolder(view);
    }

    @Override
    public void onBindViewHolder(@NonNull ReviewAdapterViewHolder holder, int position) {
        holder.mAuthorView.setText(mReviewList.get(position).getAuthor());
        holder.mContentView.setText(mReviewList.get(position).getContent());
    }

    @Override
    public int getItemCount() {
        return mReviewList == null ? 0 : mReviewList.size();
    }

    class ReviewAdapterViewHolder extends RecyclerView.ViewHolder
    {
        private final TextView mAuthorView;
        private final TextView mContentView;

        public ReviewAdapterViewHolder(View itemView) {
            super(itemView);
            mAuthorView = itemView.findViewById(R.id.tv_author);
            mContentView = itemView.findViewById(R.id.tv_content);
        }
    }
}

File: app/src/main/java/arunkbabu90/popmovies/ui/adapter/ReviewAdapter.kt
package arunkbabu90.popmovies.ui.adapter

import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.paging.PagedListAdapter
import androidx.recyclerview.widget.DiffUtil
import androidx.recyclerview.widget.RecyclerView
import arunkbabu90.popmovies.R
import arunkbabu90.popmovies.data.model.Review
import arunkbabu90.popmovies.data.repository.NetworkState
import arunkbabu90.popmovies.databinding.ItemNetworkStateBinding
import arunkbabu90.popmovies.databinding.ItemReviewBinding
import arunkbabu90.popmovies.inflate

class ReviewAdapter : PagedListAdapter<Review, RecyclerView.ViewHolder>(ReviewDiffCallback()) {
    private lateinit var binding: ItemReviewBinding
    private lateinit var networkBinding: ItemNetworkStateBinding

    val VIEW_TYPE_REVIEW = 1
    val VIEW_TYPE_NETWORK = 2

    private var networkState: NetworkState? = null

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): RecyclerView.ViewHolder {
        val inflater = LayoutInflater.from(parent.context)
        binding = ItemReviewBinding.inflate(inflater, parent, false)
        networkBinding = ItemNetworkStateBinding.bind(parent.inflate(R.layout.item_network_state))

        return if (viewType == VIEW_TYPE_REVIEW)
            ReviewViewHolder(binding.root)
        else
            NetworkStateViewHolder(networkBinding.root)
    }

    override fun onBindViewHolder(holder: RecyclerView.ViewHolder, position: Int) {
        if (getItemViewType(position) == VIEW_TYPE_REVIEW)
            (holder as ReviewViewHolder).bind(getItem(position))
        else
            (holder as NetworkStateViewHolder).bind(networkState)
    }

    override fun getItemViewType(position: Int)
        = if (hasExtraRow() && position == itemCount - 1) VIEW_TYPE_NETWORK else VIEW_TYPE_REVIEW

    override fun getItemCount() = super.getItemCount() + if (hasExtraRow()) 1 else 0

    private fun hasExtraRow(): Boolean = networkState != null && networkState != NetworkState.LOADED

    fun setNetworkState(networkState: NetworkState) {
        val prevState = this.networkState
        val hadExtraRow = hasExtraRow()
        this.networkState = networkState
        val hasExtraRow = hasExtraRow()

        if (hadExtraRow != hasExtraRow) {
            if (hadExtraRow)
                notifyItemRemoved(super.getItemCount())
            else
                notifyItemInserted(super.getItemCount())
        } else if (hasExtraRow && prevState != networkState) {
            notifyItemChanged(itemCount - 1)
        }
    }

    private inner class ReviewViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {
        fun bind(review: Review?) {
            binding.tvAuthor.text = review?.author
            binding.tvContent.text = review?.content
        }
    }

    private inner class NetworkStateViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {
        fun bind(networkState: NetworkState?) {
            if (networkState != null && networkState == NetworkState.LOADING) {
                networkBinding.itemNetworkStateProgressBar.visibility = View.VISIBLE
            } else {
                networkBinding.itemNetworkStateProgressBar.visibility = View.GONE
            }

            if (networkState != null && networkState == NetworkState.ERROR) {
                networkBinding.itemNetworkStateErrTextView.visibility = View.VISIBLE
                networkBinding.itemNetworkStateErrTextView.text = networkState.msg
            } else if (networkState != null && networkState == NetworkState.EOL) {
                networkBinding.itemNetworkStateErrTextView.visibility = View.VISIBLE
                networkBinding.itemNetworkStateErrTextView.text = networkState.msg
            } else {
                networkBinding.itemNetworkStateErrTextView.visibility = View.GONE
            }

            if (networkState != null && networkState == NetworkState.CLEAR) {
                networkBinding.itemNetworkStateErrTextView.visibility = View.VISIBLE
                networkBinding.itemNetworkStateErrTextView.text = networkState.msg
            }
        }
    }

    class ReviewDiffCallback : DiffUtil.ItemCallback<Review>() {
        override fun areItemsTheSame(oldItem: Review, newItem: Review): Boolean {
            return oldItem.id == newItem.id
        }

        override fun areContentsTheSame(oldItem: Review, newItem: Review): Boolean {
            return oldItem == newItem
        }
    }
}